### ISSUE TITLE
Advice with settings selection

### ISSUE THREAD

**Issue by elgatito** [arvidn/libtorrent#3599]
Hello,

Asking once again for your help.

Now I have few questions.
Most frequently storage uses memory to download only part of torrent file, that part is usually about 100-200mb, so in ideal - we should get pieces as quick as we can, plus, in perfect - first ones should have more priority.

1) So the question is what settings can we set in this case, to download more aggressively and get our pieces faster?

Currenly I set pieces priorities like this:
```
PrioritizePieces  Piece: 5 with priority: 7
PrioritizePieces  Piece: 6 with priority: 7
PrioritizePieces  Piece: 7 with priority: 6
PrioritizePieces  Piece: 8 with priority: 6
PrioritizePieces  Piece: 9 with priority: 5
PrioritizePieces  Piece: 10 with priority: 5
PrioritizePieces  Piece: 11 with priority: 4
PrioritizePieces  Piece: 12 with priority: 4
PrioritizePieces  Piece: 13 with priority: 3
PrioritizePieces  Piece: 14 with priority: 3
PrioritizePieces  Piece: 15 with priority: 3
PrioritizePieces  Piece: 16 with priority: 3
```

While other pieces are set priority to `0`.
2) So for the moment, when I seek the video and priorities are reorganized, I see that writes come not very sequentially, and piece with highest priority can be wrote after half of other pieces are downloaded. I understand that they should not be 100% sequential, but maybe, there are ways to make it _more_ sequential?

**Comment by arvidn** [comment]
keep in mind that the download *pipeline* may be long, and you may need to flush it when you jump your read cursor (i.e. seek). Also keep in mind that your bottleneck is likely your link to the internet, which means anything you download that you're not interested in (right away) will reduce the download rate of the things you *do* want right away.

did you see this? https://libtorrent.org/streaming.html

**Comment by elgatito** [comment]
> did you see this? https://libtorrent.org/streaming.html

I have read it few times, thanks.

> keep in mind that the download pipeline may be long, and you may need to flush it when you jump your read cursor (i.e. seek).

That is the biggest problem so far for me. We are downloading some pieces and when seek happens, we need to abort those requests and request newly prioritized pieces. 
I do set priorities for non-needed pieces to `0` and needed pieces are all with priority `>=2`. 
Then I `set_piece_deadline` to `0` for needed pieces, but now quite understanding how it works.

> you may need to flush it

How to do it?

**Comment by arvidn** [comment]
assuming you're using `set_piece_deadline()`, you can call `clear_piece_deadlines()` whenever you skip, and the set a deadline on the new high priority pieces. I don't think it will cancel the piece requests though. Maybe it should.

**Comment by arvidn** [comment]
as far as I can tell from the code, if you first clear piece deadlines, and then set new piece deadlines. All non time critical pieces (i.e. ones without deadlines) are cancelled.

**Comment by elgatito** [comment]
@arvidn I am considering usage without use of deadlines for requesting pieces. 

How it looked like:
- Every second I set priorities of needed pieces to `[1..7]`, all other set to `0` and also ...
- After priorities set, do `set_piece_deadline` for needed pieces (all with `priority > 0`).

But it looks that pieces with biggest priorities are not having faster download. 
And if I set deadline only for first N pieces - others suffer (and I see lower download speed), so it looks like mixing deadlines with usual priorities is not a great idea.

I have removed usage of deadlines and only using priorities, but on `seek` I see that I set priorities to new pieces, and expect them to be downloaded most quickly, but it takes longer than with deadlines.

So with deadlines I cannot fully use the bandwidth, and not fully understand if it cancels pending block requests and send new ones. And without deadlines I cannot prioritize first N pieces to download quick enough.

Maybe you can advice anything?

**Comment by arvidn** [comment]
> But it looks that pieces with biggest priorities are not having faster download.

Right, priority just makes peers *pick* those pieces first. If you're downloading slowly from the majority of your peers, chances are that those pieces won't complete as early as they could.

> And if I set deadline only for first N pieces - others suffer (and I see lower download speed), so it looks like mixing deadlines with usual priorities is not a great idea.

I don't recall the exact logic here, I don't think there's anything to explicitly focus bandwidth just on the peers picking deadline pieces. I would suggest you investigate what's going on here, by looking at the blocks being requested and their download progress from the various peers. (something along the lines of `TORRENT_DEBUG_STREAMING` in torrent.cpp, you could try enabling that, but it requires a terminal).

What I *do* remember is that the fast peers, that are used to pick deadline pieces, have a deeper request queue than normal, in order to get some preference.

> I have removed usage of deadlines and only using priorities, but on `seek` I see that I set priorities to new pieces, and expect them to be downloaded most quickly, but it takes longer than with deadlines.

Do you reset the priorities when you seek? You may want to try setting them to 0 even.

> So with deadlines I cannot fully use the bandwidth, and not fully understand if it cancels pending block requests and send new ones.

If you look at the pieces, you can tell: see https://libtorrent.org/reference-Core.html#get_download_queue()

**Comment by arvidn** [comment]
btw, the logic you want to trigger is this one:

   https://github.com/arvidn/libtorrent/blob/RC_1_2/src/torrent.cpp#L4579

So, it's important you remove the deadline from all pieces *first*, then seek.

**Comment by elgatito** [comment]
> Do you reset the priorities when you seek? You may want to try setting them to 0 even.

I always set priorities with a vector, setting all pieces as once, while not-needed pieces are set to 0.
I resubmit the vector once a second, with recalculation of priorities, like `0 0 0 7 6 6 5 5 5 ... 0 0 0`. 
Once the piece read and not needed - priority changes to 0.

> So, it's important you remove the deadline from all pieces first, then seek.

I was doing `clear_piece_deadlines` right when I get `seek` operation. Then usual priority updater starting and recalculates priorities for new position in the file.

First I tried this approach (Kind of pseudo code, from Go to C):
```
for (cur_piece := num_pieces - 1; cur_piece >= 0; cur_piece--) {
	if (!t.has_piece(cur_piece)) {
		t.set_piece_deadline(cur_piece, (7-piece_priority)*10, 0)
	}
}
```

Why setting from the last piece: piece with `priority = 7` is called with `deadline = 0 ms` in this case, and piece with `priority = 1` is called with `deadline = 60 ms`, and I would expect to download pieces with 0ms, and then the other ones, but, instead, the last called `set_piece_deadline` piece is downloaded first.
So with a loop `i -> [0..10]`, where I set deadlines to `0 10 20 30 40 50 ...`, the first downloaded piece would be the last. I was inspecting piece progress and could get desired behavior only with inversed loop. 

I did a simple test, starting same torrent, needed same pieces range. First case - I use priorities for needed pieces, nothing more. Second case - I use `set_piece_deadline` for needed pieces. 
In the second case I get 20% less download speed. That was with a highly seed-torrent, just to check maximum speed for each case.

The problem in my case is that:
Streaming a video would not need one reader of a file, because when you switch audio track in the file - you will have a second reader of a file, which is also prioritized (maybe even more, to allow fast read of requested pieces from a file).
And it's a real case when I have two readers of a file, let's say with ranges `[0..10]` and `[last 10 pieces]` and I need to download those range with proper priorities, so that first pieces in those ranges are downloaded in a critical manner.

**Comment by elgatito** [comment]
https://github.com/arvidn/libtorrent/blob/RC_1_1/src/torrent.cpp#L5478

In `torrent::clear_time_critical()`:
```
if (has_picker()) m_picker->set_piece_priority(i->piece, 1);
```

Piece priority is reset to 1. 
So I would need to do flow?:
1) `clear_piece_deadlines`
2) update piece priorities to 0
3) set deadlines for all the needed pieces

`torrent::remove_time_critical_pieces(std::vector<int> const& priority)`
This looks like what I need in that case, and it's triggered by `torrent::prioritize_pieces(std::vector<int> const& pieces)` , which I call once per second, right before calling `set_piece_deadline`.

**Comment by arvidn** [comment]
```
if (has_picker()) m_picker->set_piece_priority(i->piece, 1);
```

does this affect you?

do you set the priority to zero before you clear the piece deadlines? If there isn't already, there should probably be a check when setting priorities to zero to also clear the piece deadline of that piece.

**Comment by elgatito** [comment]
>  to also clear the piece deadline of that piece.

Does that mean block requests for that piece will be cancelled? 

> do you set the priority to zero before you clear the piece deadlines?

I tried the `clear_piece_deadlines` only when I've got `seek` alert. 
Currently, I set pieces priority for all the pieces (with 0 where needed), then I set deadlines for all pieces with non-0 priority.

**Comment by elgatito** [comment]
@arvidn I have tested different strategies of managing pieces and their priorities.
Tried with setting `set_piece_deadline` for all needed pieces, and `priority=0` to non-needed pieces. And these procedures are started each second.
After getting a message from user, that CPU usage is very high, I have noticed using deadlines take a lot of CPU and the speed is far from being max.

Currently I test strategy without usage of deadlines, but if I urgently need some piece - I do `set_piece_deadline(idx, 0, 0)`, and that looks better then just setting priority to 7.

Keeping the test. 
Closing the issue. Thanks
